src/com/amazon/ionschema/internal/constraint/FloatSize.kt

```diff
@@ -0,0 +1,59 @@
+package com.amazon.ionschema.internal.constraint
+
+
+import com.amazon.ion.IonFloat
+import com.amazon.ion.IonSymbol
+import com.amazon.ion.IonValue
+import com.amazon.ionschema.InvalidSchemaException
+import com.amazon.ionschema.Violation
+import com.amazon.ionschema.Violations
+import java.math.BigDecimal
+
+/**
+ * Implements the float_size constraint.
+ */
+internal class FloatSize(
+        ion: IonValue
+) : ConstraintBase(ion) {
+
+    private val validValues = setOf("float16", "float32", "float64")
+    private val maxFloat32Range: BigDecimal = BigDecimal.valueOf(Float.MAX_VALUE.toDouble())
+    private val minPosValue: BigDecimal = BigDecimal.valueOf(Float.MIN_VALUE.toDouble())
+
+
+    init {
+        if (!(ion is IonSymbol
+                        && !ion.isNullValue
+                        && validValues.contains(ion.stringValue()))) {
+            throw InvalidSchemaException("Invalid float_size constraint: $ion")
+        }
+
+        // float16 not implemented yet
+        if (ion.stringValue() == "float16") throw InvalidSchemaException("\'float16\' is not currently supported")
+    }
+
+
+    override fun validate(value: IonValue, issues: Violations) {
+        validateAs<IonFloat>(value, issues) {
+            when (ion.toString()) {
+                "float16" -> throw InvalidSchemaException("\'float16\' is not currently supported")
+                "float32" -> if (!fitsFloat32(it)) {
+                    issues.add(Violation(ion, "insufficient_float_size", "value does not fit float32 size"))
+                }
+                "float64" -> Unit
+            }
+        }
+    }
+
+
+    private fun fitsFloat32(value: IonFloat): Boolean {
```

Comment:
This function already requires that `value` be an `IonFloat`. With this in mind:
* Do we still need to transform the value into a `BigDecimal` before comparing? Or can we get away with the substantially cheaper test of seeing whether `value.doubleValue()` can be stored in a `float` and then cast back to a `double` without losing data?
* Is the `isNumericValue` test required?

Reply:
I think we get the same behaviour with the `double` approach. The reason for the use of `BigDecimal` was to avoid some of the rounding policies that Java has on float and double. 

The `isNumeric` it to distinguish the special float values `NaN` ``-Inf` `Inf` which should fit regardless of the size constraints on `float`

Reply:
> I think we get the same behaviour with the double approach. The reason for the use of BigDecimal was to avoid some of the rounding policies that Java has on float and double.

It would be good to add a comment explaining the exact behavior we need, maybe with a link to something explaining the rounding policies you mention. Without that context it's tempting to try to optimize it by avoiding the conversion.

> The isNumeric it to distinguish the special float values NaN ``-Inf Inf` which should fit regardless of the size constraints on `float`

Ahh, got it! I thought that `isNumericValue` was on the `IonValue` interface (to tell if the value is an `int`, `float`, or `decimal`), but I see now that it's actually [a method on `IonFloat`](https://www.javadoc.io/doc/com.amazon.ion/ion-java/latest/com/amazon/ion/IonFloat.html#isNumericValue--). That might also be nice to explain in a comment.

Reply:
Would a comment like the following do? 

```
/**  
 * Implements the float_size  *
                  * float32 validation relies on the precision of constants Float.MAX_VALUE.toDouble() and Float.MIN_VALUE.toDouble().
                  * Client code working with floating points should use .toDouble()'s precision for FloatSize validation to work.
                  */


```

Reply:
That comment makes sense. I guess I was hoping to see an example of a place where `double`->`float`->`double` would lose precision in a way that keeps it from being a good test for `fitsInFloat32`.

Reply:
This has an example, no exactly the order in your comment. 
https://stackoverflow.com/questions/916081/convert-float-to-double-without-losing-precision